docs/TESTPLAN.md

```diff
@@ -0,0 +1,68 @@
+# Test Plan
+
+This document describes the testing architecture, goals and current status for
+the AWS Toolkit for VSCode.
+
+## Test goals
+
+In order of priority:
+
+1. Fast feedback from making a code change to seeing a result
+2. High confidence in releasing the Toolkit
+3. Avoid regressions
+
+Ratio of unit to integ tests: 90% unit tests, 10% system/acceptance tests.
+
+## Test categories
+
+The test suite has two categories of tests:
+
+-   Unit Tests: **fast** tests
+    -   Live in `src/test/`
+    -   The `vscode` API is available, but a full VSCode instance is not started.
+    -   The Toolkit code is invoked as a library, not as an extension loaded in VSCode.
+    -   Call functions and create objects directly.
+    -   May mock state where needed, though this is discouraged in favor of "fake" data/objects/files.
+    -   May use the filesystem.
+    -   Main property is that [the test is fast](https://pycon-2012-notes.readthedocs.io/en/latest/fast_tests_slow_tests.html).
+-   Integration Tests: **slow** tests
+    -   Live in `src/integrationTest/`
+    -   Use a full instance of VSCode.
+    -   Global state is shared across tests, thus later tests can be polluted by earlier tests.
```

Comment:
consider replacing "can be polluted by earlier tests" with "run the risk of being polluted by earlier tests"
(impact vs permission)

Reply:
I don't think this is unique to our integration tests? May be worth a cleanup pass in the unit tests (although we typically figure it out quickly enough).

Reply:
cleanup pass ?

Reply:
We've definitely had unit tests write to global state before, and we only pick it up when it actually causes issues. I'm saying it's not unique to integration tests, but if we want it to be unique, we should go through our existing unit tests with a fine-toothed comb to see what we can find (and either correct or convert to integration tests)